jobs/ror/DownloadAndExtractRorDataset.php

```diff
@@ -0,0 +1,226 @@
+                throw new Exception('Failed to determine file size');
+            }
+
+            $this->performChunkDownload(
+                $pathZipDir,
+                $pathZipFile,
+                $this->downloadUrl,
+                $chunkDir,
+                $fileSize
+            );
+
+        } catch (Throwable $e) {
+            UpdateRorRegistryDataset::cleanup([$pathZipFile, $pathZipDir]);
+            UpdateRorRegistryDataset::log(
+                "RoR dataset downloading in chunk and extract main job have failed: {$e->getMessage()}",
+                $this->scheduledTaskLogFilesPath,
+                ScheduledTaskHelper::SCHEDULED_TASK_MESSAGE_TYPE_ERROR
+            );
+            throw $e;
+        }
+    }
+
+    protected function performChunkDownload(
+        string $pathZipDir,
+        string $pathZipFile,
+        string $downloadUrl,
+        string $chunkDir,
+        int $fileSize
+    )
+    {
+        try {
+
+            if (!$this->fileManager->mkdir($chunkDir)) {
+                throw new Exception('Failed to create chunk directory');
+            }
+
+            $chunkSize = 1024 * 1024 * 3 ; // 3 MB
+            $numChunks = ceil($fileSize / $chunkSize);
+            $jobs = [];
+
+            for ($i = 0; $i < $numChunks; $i++) {
+                $startByte = $i * $chunkSize;
+                $endByte = min(($i + 1) * $chunkSize - 1, $fileSize - 1);
+                $chunkFile = $chunkDir . "/chunk_{$i}";
+                $jobs[] = new DownloadRoRDatasetInChunks(
+                    $downloadUrl,
+                    $startByte,
+                    $endByte,
+                    $chunkFile,
+                    $pathZipDir,
+                    $this->csvNameContains,
+                    $this->scheduledTaskLogFilesPath
+                );
+            }
+
+            // we need to get thses in local variables to pass into the batch closures(e.g. then, catch, etc)
+            // as the serialization of $this in closure is buggy and can lead to issues
+            $prefix = $this->prefix;
+            $csvNameContains = $this->csvNameContains;
+            $scheduledTaskLogFilesPath = $this->scheduledTaskLogFilesPath;
+
+            // Dispatch batch
+            Bus::batch($jobs)
+                ->then(function (Batch $batch) use ($pathZipDir, $pathZipFile, $chunkDir, $fileSize, $chunkSize, $prefix, $csvNameContains, $scheduledTaskLogFilesPath) {
+                    // after successfully completing all the batch jobs, 
+                    // first we need to verify if all chunks downloaded
+                    $chunks = glob($chunkDir . '/chunk_*');
+
+                    if (count($chunks) !== (int) ceil($fileSize / $chunkSize)) {
+                        // not all chunks downloaded
+                        // we will not attempt any more chunk downloads and cancel batch process
+                        // and fall back to sync download
+                        $batch->cancel();
```

Comment:
Is this needed, because if `then` is called, all jobs were successfully completed ?